src/rpc/auxpow.cpp

```diff
@@ -150,19 +139,20 @@ static UniValue AuxMiningCreateBlock(const CScript& scriptPubKey)
     return result;
 }
 
-static bool AuxMiningSubmitBlock(const std::string& hashHex, const std::string& auxpowHex)
+static UniValue AuxMiningSubmitBlock(const std::string& hashHex, const std::string& auxpowHex)
 {
-
     AuxMiningCheck();
-    LOCK(cs_auxblockCache);
+    LOCK(cs_auxpowrpc);
 
     uint256 hash;
     hash.SetHex(hashHex);
 
-    const std::map<uint256, CBlock*>::iterator mit = mapNewBlock.find(hash);
-    if (mit == mapNewBlock.end())
+    std::shared_ptr<CBlock> pblock;
+    bool maybe_block = auxBlockCache.Get(hash, pblock);
+    if (!maybe_block) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "block hash unknown");
-    CBlock& block = *mit->second;
+    }
+    CBlock& block = *pblock;
```

Comment:
I'm struggling a bit to figure out the shared pointer lifecycle here; we should rely on cache expiration/clearing to clean this up here, right? No need for an explicit `reset()` in this codepath?

Reply:
Oh right, copy on assignment here. Sorry, I should have re-read the class operators before asking. This all makes sense. I was reading the code closely to make sure we handled the lifecycle correctly so this codepath jumped out at me.